clippy_lints/src/zombie_processes.rs

```diff
@@ -0,0 +1,253 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::visitors::for_each_local_use_after_expr;
+use clippy_utils::{fn_def_id, get_enclosing_block, match_any_def_paths, match_def_path, paths};
+use rustc_ast::Mutability;
+use rustc_hir::def::DefKind;
+use rustc_hir::intravisit::{walk_expr, Visitor};
+use rustc_hir::{Expr, ExprKind, HirId, Local, Node, PatKind, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{sym, Span};
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for code that spawns a process but never calls `wait()` on the child.
+    ///
+    /// ### Why is this bad?
+    /// As explained in the [standard library documentation](https://doc.rust-lang.org/stable/std/process/struct.Child.html#warning),
+    /// calling `wait()` is necessary on Unix platforms to properly release all OS resources associated with the process.
+    /// Not doing so will effectively leak process IDs and/or other limited global resources,
+    /// which can eventually lead to resource exhaustion, so it's recommended to call `wait()` in long-running applications.
+    /// Such processes are called "zombie processes".
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::process::Command;
+    ///
+    /// let _child = Command::new("ls").spawn().expect("failed to execute child");
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::process::Command;
+    ///
+    /// let mut child = Command::new("ls").spawn().expect("failed to execute child");
+    /// child.wait().expect("failed to wait on child");
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub ZOMBIE_PROCESSES,
+    suspicious,
+    "not waiting on a spawned child process"
+}
+declare_lint_pass!(ZombieProcesses => [ZOMBIE_PROCESSES]);
+
+fn emit_lint(cx: &LateContext<'_>, span: Span) {
+    span_lint_and_then(
+        cx,
+        ZOMBIE_PROCESSES,
+        span,
+        "spawned process is never `wait()`-ed on and leaves behind a zombie process",
+        |diag| {
+            diag.help("consider calling `.wait()`")
+                .note("also see https://doc.rust-lang.org/stable/std/process/struct.Child.html#warning");
```

Comment:
This is a bit noisy imo:
```suggestion
            diag.help("see https://doc.rust-lang.org/stable/std/process/struct.Child.html#warning");
```
Ideally it'd also have a suggestion, but are there any cases where adding it to the end of the span wouldn't work?

Reply:
Yea, I guess we can have a suggestion for `let _ = <expr that spawns child>;` and `<expr that spawns child>;` specifically. But probably not for when it's assigned to a binding, since it's used elsewhere and finding an appropriate place to put the call seems tricky.

It should probably also be `MaybeIncorrect`. Maybe the user wants `kill()` instead 🤷